ocaml/gencert/lib.ml

```diff
@@ -32,3 +32,115 @@ let get_management_ip_addr ~dbg =
       let addrs = List.filter (fun addr -> String.sub addr 0 4 <> "fe80") addrs in
       Some (List.hd addrs)
   with _ -> None
+
+open Api_errors
+open Rresult
+
+type t_certificate = Leaf | Chain
+
+let validate_private_key pkcs8_private_key =
+  let ensure_key_length = function
+  | `RSA priv ->
+    let length = Mirage_crypto_pk.Rsa.priv_bits priv in
+    if length < 2048 || length > 4096 then
+      Error (`Msg (server_certificate_key_rsa_length_not_supported,
+        [Int.to_string length]))
+    else
+      Ok (`RSA priv)
+  in
+
+  let raw_pem = Cstruct.of_string pkcs8_private_key in
+  X509.Private_key.decode_pem raw_pem
+  |> R.reword_error
+      (fun (`Msg err_msg) ->
+        let unknown_algorithm = "Unknown algorithm " in
+        if Astring.String.is_prefix ~affix:"multi-prime RSA" err_msg then
+          `Msg (server_certificate_key_rsa_multi_not_supported, [])
+        else if Astring.String.is_prefix ~affix:unknown_algorithm err_msg then
+          `Msg (server_certificate_key_algorithm_not_supported,
+            [Astring.String.with_range ~first:(String.length unknown_algorithm) err_msg])
+        else
+          `Msg (server_certificate_key_invalid, []))
+  >>= ensure_key_length
+
+let validate_certificate kind pem now private_key =
+  let ensure_keys_match private_key certificate =
+    let public_key = X509.Certificate.public_key certificate in
+    match public_key, private_key with
+    | `RSA pub, `RSA priv when pub = Mirage_crypto_pk.Rsa.pub_of_priv priv ->
+        Ok certificate
+    | _ ->
+        Error (`Msg (server_certificate_key_mismatch, []))
+  in
+
+  let ensure_validity ~time certificate =
+    let to_string = Ptime.to_rfc3339 ~tz_offset_s:0 in
+    let not_before, not_after = X509.Certificate.validity certificate in
+
+    if Ptime.is_earlier time not_before then
+      Error (`Msg (server_certificate_not_valid_yet,
+        [(to_string time); (to_string not_before)]))
+    else if Ptime.is_later time not_after then
+      Error (`Msg (server_certificate_expired,
+          [(to_string time); (to_string not_after)]))
+    else
+      Ok certificate
+  in
+
+  let ensure_sha256_signature_algorithm certificate =
+    match X509.Certificate.signature_algorithm certificate with
+    | Some (_, `SHA256) ->
+        Ok certificate
+    | _ ->
+        Error (`Msg (server_certificate_signature_not_supported, []))
+  in
+
+  let raw_pem = Cstruct.of_string pem in
+  (match kind with
```

Comment:
Why do we need `(` here?

Reply:
It's not needed, probably a leftover when it wasn't the last statement in the function.